Make translation of names injective

Lysxia opened the pull request:
Closes #1107

- There are some `snake_case` conversions remaining somewhere in `creusot/src/backend/dependency.rs`, are they related?
- The `why3session.xml` files need to be updated too. Is there a tool for that?

jhjourdan commented:
>     * There are some `snake_case` conversions remaining somewhere in `creusot/src/backend/dependency.rs`, are they related?

Yes, I think they should also use a new encoding. The difference here is that they have to start with a lower case letter.

> 
>     * The `why3session.xml` files need to be updated too. Is there a tool for that?

If you run `cargo test --test why3`, you tell Why3 to replay all the proofs. It should be able to recover the proofs even though the identifiers have changed. If the repay succeeds, then it should write "obsolete" in orange, and it overrides the session files on the fly.

In addition, please run the `fmt` script to normalize the formatting of Rust code. If you find this frustrating, please complain to @xldenis, we may perhaps make him change his mind ;)

xldenis commented:
> In addition, please run the fmt script to normalize the formatting of Rust code. If you find this frustrating, please complain to @xldenis, we may perhaps make him change his mind ;)

Nice try! but I would recommend just setting up format-on-save so you no longer need to think about it :)

xldenis commented:
> The why3session.xml files need to be updated too. Is there a tool for that?

Unfortunately, apart from `grep` no.

xldenis commented:
> There are some snake_case conversions remaining somewhere in creusot/src/backend/dependency.rs, are they related?

Yep!

xldenis commented on creusot/src/util.rs:
```diff
@@ -248,31 +248,75 @@ pub(crate) fn module_name(tcx: TyCtxt, def_id: DefId) -> Symbol {
+    // Escape initial underscore
+    if chars.peek() == Some(&'_') {
+        let _ = chars.next();
+        dest.push('q');
+        dest.push('y');
+        dest.push('9');
+        dest.push('5');
+        dest.push('z');
```
You should be able to use `.extend("qy95z")` here.

Lysxia replied:
I found `.push_str()` which is similar. Do you have any preference vs `.extend()`?

xldenis replied:
thats good. I forgot that `String` has a weird api there for whatever reason.

xldenis commented on creusot/src/util.rs:
```diff
@@ -248,31 +248,75 @@ pub(crate) fn module_name(tcx: TyCtxt, def_id: DefId) -> Symbol {
+        if c == '_' {
+            match chars.peek() {
+                None | Some('_') => {
+                    dest.push('q');
+                    dest.push('y');
+                    dest.push('9');
+                    dest.push('5');
+                    dest.push('z');
```
same,

xldenis commented on creusot/src/util.rs:
```diff
@@ -248,31 +248,75 @@ pub(crate) fn module_name(tcx: TyCtxt, def_id: DefId) -> Symbol {
+        } else if c.is_ascii_alphanumeric() && !is_qy {
+            dest.push(c);
+        } else {
+            dest.push('q');
+            dest.push('y');
+            // convert code point c to decimal
+            for d in format!("{}", c as u32).chars() {
+                dest.push(d);
+            }
```
`.extend(format!("qy{}z", c as u32))`

Lysxia commented on creusot/src/backend/dependency.rs:
```diff
@@ -172,22 +171,17 @@ impl<'tcx> Dependency<'tcx> {
+                    item_symb(tcx, *def_id, rustc_hir::def::Namespace::TypeNS)
+                }
+                _ => Symbol::intern("debug_ty_name"),
+            },
```
Here the identifiers already seem to have been initial-lowercased (as far as the test suite can tell)

xldenis commented:
@Lysxia if you look at the `why3` test suite (`cargo test --test why3`
) you'll see some files which contain syntax errors. If you can fix those, someone else can help you update the obsolete proof sessions if needed. You can avoid running the provers to find syntax issues faster using `cargo test --test why3 -- --replay=none`

Lysxia commented:
I fixed the syntax errors. Because the names of the theories changed, why3 sees a lot of detached goals so the sessions are quite broken. We can probably generate a sed script to do the renaming in the session files, unless someone has a better idea.

I removed all uses of `to_camel_case` and `to_snake_case`, but there are still some `to_lowercase` or `make_ascii_lowercase` that make me a bit uneasy.

xldenis commented:
If you run `cargo test --test why3` it should fix most of the files automatically, but yea otherwise a sed may be the most efficient manner of fixing

Lysxia commented:
I patched the Why3 sessions and there are only 4 failing ones left on my machine:
- `list_reversal_lasso`, trying to `remove` no longer existing assumptions
- `red_black_tree`, timeouts
- `selection_sort_generic`, timeouts
- `sparse_array`, broken `pose`

jhjourdan commented:
Seems like the only problem is with `sparse_array`, with some detached goal that need to be cleaned. I'm hitting the "Enable auto-merge" button so that once you'll have fixed that, this will be merged.
